### PR TITLE
drivers/gpio: Add support for GRLIB GRGPIO2

### PR DESCRIPTION
**Pull request by tbr-tt**
This adds support for the GRLIB GRGPIO2 controller used in LEON and NOEL-V systems.

Also activates GPIO support in the GR716A SOC and the `gr716a_mini` board.

### CODE REVIEW

**Comment by tbr-tt** [comment]
The compliance check error (checkpatch.pl) says:
```
-:158: WARNING:LINE_SPACING: Missing a blank line after declarations
#158: FILE: drivers/gpio/gpio_grgpio2.c:27:
+	struct gpio_driver_config common;
+	volatile struct grgpio_regs *regs;
```

That is for the structure defined as:
```
struct cfg {
	struct gpio_driver_config common;
	volatile struct grgpio_regs *regs;
	int interrupt;
};
```

Is there a recommendation on how to update the structure here?

**Comment by MaureenHelm** [comment]
@julius-barendt can you take a look please?

**Review by julius-barendt — CHANGES_REQUESTED** [review]
Sequence registers are missing, but I don't think they are needed as of now as they aren't used. But please update the dtsi so the reg property gets the correct size

**Comment by julius-barendt** [comment]
> The compliance check error (checkpatch.pl) says:
> 
> ```
> -:158: WARNING:LINE_SPACING: Missing a blank line after declarations
> #158: FILE: drivers/gpio/gpio_grgpio2.c:27:
> +	struct gpio_driver_config common;
> +	volatile struct grgpio_regs *regs;
> ```
> 
> That is for the structure defined as:
> 
> ```
> struct cfg {
> 	struct gpio_driver_config common;
> 	volatile struct grgpio_regs *regs;
> 	int interrupt;
> };
> ```
> 
> Is there a recommendation on how to update the structure here?

Looks like a bug in the script related to the volatile property. Inserting a blank line after common will fix it. But I don't know if there is a better way

File: dts/sparc/gaisler/gr716a.dtsi [hunk 1]
```diff
@@ -107,5 +108,23 @@
 			#size-cells = <0>;
 			status = "disabled";
 		};
+
+		gpio0: gpio@8030c000 {
+			interrupts = <17>;
+			reg = <0x8030c000 0x100>;
```
[the remark this answers is not in the document]
Comment (tbr-tt) [reply dts/sparc/gaisler/gr716a.dtsi]: This is updated in the 2024-04-17 push. Thanks!

File: dts/sparc/gaisler/gr716a.dtsi [hunk 2]
```diff
@@ -107,5 +108,23 @@
...
+			compatible = "gaisler,grgpio";
+			gpio-controller;
+			#gpio-cells = <2>;
+			status = "disabled";
+		};
+
+		gpio1: gpio@8030d000 {
+			interrupts = <16>;
+			reg = <0x8030d000 0x100>;
```
[the remark this answers is not in the document]
Comment (tbr-tt) [reply dts/sparc/gaisler/gr716a.dtsi]: Also updated now.

**Review by julius-barendt — DISMISSED** [review]

File: drivers/gpio/gpio_grgpio2.c [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	};								\
+	static struct data data_##n;					\
+									\
+	DEVICE_DT_INST_DEFINE(n,					\
+			    grgpio_init,				\
+			    NULL,					\
+			    &data_##n,					\
+			    &cfg_##n,					\
+			    POST_KERNEL,				\
+			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_grgpio2.c:300]:
```suggestion
			    CONFIG_GPIO_INIT_PRIORITY,		\
```

File: drivers/gpio/gpio_grgpio2.c [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+	.port_set_masked_raw            = port_set_masked_raw,
+	.port_set_bits_raw              = port_set_bits_raw,
+	.port_clear_bits_raw            = port_clear_bits_raw,
+	.port_toggle_bits               = port_toggle_bits,
+	.pin_interrupt_configure        = pin_interrupt_configure,
+	.manage_callback                = manage_callback,
+	.get_pending_int                = get_pending_int,
+};
+
+#define THE_INIT(n)							\
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_grgpio2.c:284]:
```suggestion
#define GRGPIO_INIT(n)							\
```

File: dts/bindings/gpio/gaisler,grgpio.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+description: GRLIB GRGPIO - General Purpose I/O Port
```
Comment (henrikbrixandersen) [inline dts/bindings/gpio/gaisler,grgpio.yaml:1]:
```suggestion
description: Gaisler GRLIB GRGPIO - General Purpose I/O Port
```

File: soc/gaisler/gr716a/Kconfig.defconfig [hunk 6]
```diff
@@ -23,4 +23,8 @@ config SPI_GRLIB_SPIMCTRL
 
 endif
 
+config GPIO_GRGPIO2
+	default y
+	depends on GPIO
+
```
Comment (henrikbrixandersen) [inline soc/gaisler/gr716a/Kconfig.defconfig:29]: This should not be needed.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_grgpio2.c (hunk 3 above)
Comment (tbr-tt) [reply drivers/gpio/gpio_grgpio2.c]: Updated according to your recommendation.

File: drivers/gpio/gpio_grgpio2.c (hunk 4 above)
Comment (tbr-tt) [reply drivers/gpio/gpio_grgpio2.c]: Updated according to your recommendation.

File: dts/bindings/gpio/gaisler,grgpio.yaml (hunk 5 above)
Comment (tbr-tt) [reply dts/bindings/gpio/gaisler,grgpio.yaml]: Updated according to your recommendation.

File: soc/gaisler/gr716a/Kconfig.defconfig (hunk 6 above)
Comment (tbr-tt) [reply soc/gaisler/gr716a/Kconfig.defconfig]: It has been removed. Thanks!

**Review by henrikbrixandersen — APPROVED** [review]

**Review by julius-barendt — APPROVED** [review]

**Pull request merged**